src/lib.rs

```diff
@@ -0,0 +1,341 @@
+extern crate chrono;
+extern crate oauth2;
+extern crate reqwest;
+extern crate url;
+#[macro_use]
+extern crate serde_derive;
+extern crate serde;
+extern crate tiny_http;
+#[macro_use]
+extern crate failure;
+
+use chrono::NaiveDate;
+use oauth2::{AuthType, Config as OAuth2Config};
+use reqwest::header::{Authorization, Bearer, Headers, UserAgent};
+use reqwest::Method;
+
+pub mod errors;
+
+use errors::Error;
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Token(oauth2::Token);
+
+pub struct FitbitClient {
+    client: reqwest::Client,
+    base: url::Url,
+}
+
+pub enum DateQuery {
+    ForDate(NaiveDate),
+    PeriodicSince(NaiveDate, Period),
+    Range(NaiveDate, NaiveDate),
+}
+
+/// UserProfile is a partial serialization struct of the Fitbit API profile. See:
+/// https://dev.fitbit.com/build/reference/web-api/user/
+#[derive(Serialize, Deserialize, Debug)]
+pub struct UserProfile {
+    age: i64,
+    #[serde(rename = "offsetFromUTCMillis")]
+    utc_offset: i64,
+    // TODO: lots of fields...
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct UserProfileResult {
+    user: UserProfile,
+}
+
+pub trait User {
+    fn get_profile(&self) -> Result<UserProfileResult, Error>;
+}
+
+impl User for FitbitClient {
+    fn get_profile(&self) -> Result<UserProfileResult, Error> {
+        let url = self.base.join("user/-/profile.json")?;
+        Ok(self
```

Comment:
I think this code will be repeated in a lot of places. Leaving it "stringly"-typed we could extract a helper method like

```rust
fn get_profile(&self) -> Result<???> {
    let url = self.base.join(...)
    self.client.do_get(url)
}

impl FitbitClient {
    fn do_get(&self, path: &str) -> Result<String> {
        let url = self.base.join(&path)?;
        self.client
            .request(Method::Get, url)
            .send()
            .and_then(|mut r| r.text())
    }
}
```

With the strongly-typed API, I'd have to think about how to do this method extraction. Do you have any ideas?

Reply:
yeah, a helper would be good here; (hand-waving) something like `do_get<T>(url: &str) -> Result<T, Error>`.  There must be a way to make this generic plumbing for what eventually is serde `Deserialize`.